ext/semian/sliding_window.c

```diff
@@ -0,0 +1,212 @@
+  }
+
+  return retval;
+}
+
+VALUE semian_simple_sliding_window_last(VALUE self) {
+  printf("[DEBUG] semian_simple_sliding_window_last\n");
+
+  semian_simple_sliding_window_t *res;
+  TypedData_Get_Struct(self, semian_simple_sliding_window_t, &semian_simple_sliding_window_type, res);
+  semian_simple_sliding_window_shared_t *window = get_window(res->key);
+
+  int index = (window->start + window->length - 1) % window->max_size;
+  printf("[DEBUG]   index:%d last:%d\n", index, window->data[index]);
+
+  return RB_INT2NUM(window->data[index]);
+}
+
+VALUE semian_simple_sliding_window_clear(VALUE self) {
+  printf("[DEBUG] semian_simple_sliding_window_clear\n");
+
+  semian_simple_sliding_window_t *res;
+  TypedData_Get_Struct(self, semian_simple_sliding_window_t, &semian_simple_sliding_window_type, res);
+  semian_simple_sliding_window_shared_t *window = get_window(res->key);
+
+  window->length = 0;
+  window->start = 0;
+  window->end = 0;
+
+  return self;
+}
+
+VALUE semian_simple_sliding_window_reject(VALUE self) {
+  printf("[DEBUG] semian_simple_sliding_window_reject\n");
+
+  rb_need_block();
+
+  semian_simple_sliding_window_t *res;
+  TypedData_Get_Struct(self, semian_simple_sliding_window_t, &semian_simple_sliding_window_type, res);
+  semian_simple_sliding_window_shared_t *window = get_window(res->key);
+
+  // Store these values because we're going to be modifying the buffer.
+  int start = window->start;
+  int length = window->length;
+
+  int cleared = 0;
+  for (int i = 0; i < length; ++i) {
+    int index = (start + i) % length;
+    int value = window->data[index];
+    printf("[DEBUG]   i:%d index: %d value:%d max_size:%d length:%d start:%d end:%d\n", i, index, value, window->max_size, window->length, window->start, window->end);
+    VALUE y = rb_yield(RB_INT2NUM(value));
+    if (RTEST(y)) {
+      if (cleared++ != i) {
+        rb_raise(rb_eArgError, "reject! must delete monotonically");
+      }
+      window->start = (window->start + 1) % window->length;
+      window->length--;
+      printf("[DEBUG]   Removed index:%d (val:%d)\n", i, value);
+    }
+  }
+
+  return self;
+}
+
+VALUE semian_simple_sliding_window_push(VALUE self, VALUE value) {
+  printf("[DEBUG] semian_simple_sliding_window_push\n");
+
+  semian_simple_sliding_window_t *res;
+  TypedData_Get_Struct(self, semian_simple_sliding_window_t, &semian_simple_sliding_window_type, res);
+
+  semian_simple_sliding_window_shared_t *window = get_window(res->key);
+  if (window->length == window->max_size) {
+    window->length--;
+    window->start = (window->start + 1) % window->max_size;
+  }
+
+  const int index = window->end;
+  window->length++;
+  window->data[index] = RB_NUM2INT(value);
+  window->end = (window->end + 1) % window->max_size;
+
+  printf("[DEBUG]   Pushed val:%d index:%d max_size:%d length:%d start:%d end:%d\n", RB_NUM2INT(value), index, window->max_size, window->length, window->start, window->end);
+
+  return self;
+}
```

Comment:
Circular buffers 👏 🎉

Reply:
Yeah, it's a little complex working with shared memory because we can't have dynamic data structures. So the window size has to be fixed, and I used a circular buffer to work around the memory management issues.